integration-tests/src/test/java/com/datastax/oss/driver/core/TCPFlowControlIT.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.core;
+
+import static com.datastax.oss.simulacron.common.stubbing.PrimeDsl.noRows;
+import static com.datastax.oss.simulacron.common.stubbing.PrimeDsl.when;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import com.codahale.metrics.Gauge;
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.DriverTimeoutException;
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.core.data.ByteUtils;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metrics.DefaultNodeMetric;
+import com.datastax.oss.driver.api.testinfra.session.SessionUtils;
+import com.datastax.oss.driver.api.testinfra.simulacron.SimulacronRule;
+import com.datastax.oss.driver.categories.IsolatedTests;
+import com.datastax.oss.driver.internal.core.session.DefaultSession;
+import com.datastax.oss.simulacron.common.cluster.ClusterSpec;
+import com.datastax.oss.simulacron.common.request.Query;
+import com.datastax.oss.simulacron.server.BoundNode;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import org.awaitility.Awaitility;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(IsolatedTests.class)
+public class TCPFlowControlIT {
```

Comment:
I wouldn't mind keep these tests – they are very nicely designed btw – but:

1. We are testing rather Netty than the driver here;
2. I wonder how deterministic this test really is? I think it could fail in many ways depending on how the underlying socket is implemented. I'm a bit worried that this test will be flaky forever.

Reply:
1. I don't have strong opinion here. But since the netty is the core of our driver, sometimes it may be beneficial to test the feature of netty that we strongly rely on. Should the netty change the behavior of TCP buffering, we will notice it immediately. I see it a bit similar to performance tests that are we are running if the netty version is incremented. Here, we are not testing performance directly but another important aspect like buffering (it can have impact on performance as well). So maybe it is beneficial to have it automated via IT?
2. I run them a lot of times and they are constantly passing. The tests are also green on Jenkins. I didn't observe the flaky behavior of those ITs.

Reply:
Good arguments. Let's have the test then, and if we ever notice any flakiness we can revisit this later.